lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/file/MultipartUploadComposerDAOImpl.java

```diff
@@ -0,0 +1,172 @@
+
+	@Autowired
+	private DBOBasicDao basicDao;
+
+	private RowMapper<DBOMultipartUploadComposerPartState> rowMapper = (rs, rowNum) -> {
+		DBOMultipartUploadComposerPartState dbo = new DBOMultipartUploadComposerPartState();
+		dbo.setUploadId(rs.getLong(COL_MULTIPART_COMPOSER_PART_UPLOAD_ID));
+		dbo.setPartRangeUpperBound(rs.getLong(COL_MULTIPART_COMPOSER_PART_RANGE_UPPER_BOUND));
+		dbo.setPartRangeLowerBound(rs.getLong(COL_MULTIPART_COMPOSER_PART_RANGE_LOWER_BOUND));
+		return dbo;
+	};
+
+	private RowMapper<PartRange> partRangeRowMapper = (rs, rowNum) -> {
+		PartRange dto = new PartRange();
+		dto.setUpperBound(rs.getLong(COL_MULTIPART_COMPOSER_PART_RANGE_UPPER_BOUND));
+		dto.setLowerBound(rs.getLong(COL_MULTIPART_COMPOSER_PART_RANGE_LOWER_BOUND));
+		return dto;
+	};
+
+	@Override
+	public void addPartToUpload(String uploadId, long lowerBound, long upperBound) {
+		ValidateArgument.required(uploadId, "UploadId");
+		// update the etag of the master row.
+		updateEtag(uploadId);
+		// update the part state
+		DBOMultipartUploadComposerPartState partState = new DBOMultipartUploadComposerPartState();
+		partState.setUploadId(Long.parseLong(uploadId));
+		partState.setPartRangeLowerBound(lowerBound);
+		partState.setPartRangeUpperBound(upperBound);
+		basicDao.createOrUpdate(partState);
+	}
+
+	@Override
+	public List<DBOMultipartUploadComposerPartState> getAddedParts(Long uploadId) {
+		MapSqlParameterSource param = new MapSqlParameterSource().addValue(PARAM_UPLOAD_ID, uploadId);
+		return namedJdbcTemplate.query(SQL_SELECT_PARTS_BY_UPLOAD_ID, param, rowMapper);
+	}
+
+	@Override
+	public List<PartRange> getAddedPartRangesForUpdate(Long uploadId, Long lowerBound, Long upperBound) {
+		MapSqlParameterSource param = new MapSqlParameterSource()
+				.addValue(PARAM_UPLOAD_ID, uploadId)
+				.addValue(PARAM_LOWER_BOUND, lowerBound)
+				.addValue(PARAM_UPPER_BOUND, upperBound);
+		return namedJdbcTemplate.query(SQL_SELECT_PARTS_IN_RANGE + FOR_UPDATE, param, partRangeRowMapper);
+	}
+
+	@Override
+	public void deletePartsInRange(String uploadId, long lowerBound, long upperBound) {
+		namedJdbcTemplate.update(SQL_DELETE_PARTS_IN_RANGE, new MapSqlParameterSource()
+				.addValue(PARAM_UPLOAD_ID, uploadId)
+				.addValue(PARAM_LOWER_BOUND, lowerBound)
+				.addValue(PARAM_UPPER_BOUND, upperBound));
+	}
+
+	@Override
+	public CompositeMultipartUploadStatus setUploadComplete(String uploadId, String fileHandleId) {
+		ValidateArgument.required(uploadId, "UploadId");
+		ValidateArgument.required(fileHandleId, "FileHandleId");
+		String newEtag = UUID.randomUUID().toString();
+		MultipartUploadState state = MultipartUploadState.COMPLETED;
+
+		MapSqlParameterSource updateParam = new MapSqlParameterSource()
+				.addValue(PARAM_FILE_HANDLE_ID, fileHandleId)
+				.addValue(PARAM_ETAG, newEtag)
+				.addValue(PARAM_MULTIPART_STATE, state.name())
+				.addValue(PARAM_UPLOAD_ID, uploadId);
+
+		namedJdbcTemplate.update(SQL_SET_COMPLETE, updateParam);

```

Comment:
move this set complete to manager (the other dao may already be taking care of this)